Fixed stop time availablity time frame

WuestFlorian opened the pull request:
Changed the time frame to calculate the `availabiltiy of acceptable stop time updates` metric to use the first and last provided update for each trip and stop combination.

Previously, it was the first and last update published across all trip and stop combinations, but this was lost somewhere. As our test feed fulfills the required two updates per minute at no point, the result stayed the same (0.0%) as therefore was not noticed.

In the official definition, the time frame is defined as the `trip start time` up to the `actual arrival time`. In my opinion, this only makes sense if one observes the published stop time updates for this entire time, as otherwise many slots would be marked as having not enough updates, even if they have and we just did not capture them.

Therefore, I propose to use the time frame as the range we have stop time updates in, as this should be a good estimate for the entire trip.

WuestFlorian committed:
Fixed stop time availablity time frame

Changed stop time availability time frame to time frame of provided stop time updates, added a debug logging option

WuestFlorian committed:
Remove testing code

maexled commented on src/realtime_metrics/main.py:
```diff
@@ -363,7 +365,7 @@ def availability_acceptable_stop_time_updates(stop_time_updates: list[tuple[Trip
 
         # get time in minutes
-        time_in_minutes = int(trip_update.timestamp.replace(tzinfo=timezone.utc).timestamp() / 60)
+        time_in_minutes = int(trip_update.timestamp.replace(tzinfo=timezone.utc).timestamp() / 60) * 60
```
/60 * 60? So this are seconds now? Rename the variable please if so.

WuestFlorian replied:
I kept the timestamp in seconds, as timestamps are defined this way. This operations removes the part responsible for individual seconds and thereby round down to the last minute. I renamed the variable and adapted the comment to make this more clear.

maexled commented on src/realtime_metrics/main.py:
```diff
@@ -79,6 +76,11 @@ def run_analysis():
+        if len(trip_updates) == 0:
+            continue
+        trip_updates.sort(key=lambda u: u[0].timestamp.replace(tzinfo=timezone.utc).timestamp())
+        time_frame_start = int(trip_updates[0][0].timestamp.replace(tzinfo=timezone.utc).timestamp() / 60) * 60
+        time_frame_end = int(trip_updates[-1][0].timestamp.replace(tzinfo=timezone.utc).timestamp() / 60) * 60
```
```suggestion
        time_frame_start = int(trip_updates[0][0].timestamp.replace(tzinfo=timezone.utc).timestamp())
        time_frame_end = int(trip_updates[-1][0].timestamp.replace(tzinfo=timezone.utc).timestamp())
```

WuestFlorian replied:
This would remove the rounding down to the last minute. While it is not necessary to have the timeframe already rounded, it make debugging easier, so I would keep it.

1Maxnet1 replied:
Did you consider to use https://www.w3schools.com/python/ref_func_round.asp ? This int(x / 60) * 60 is a litlle bit confusing imho.

WuestFlorian replied:
I found a function to drop the seconds on the `datetime` object directly. It should be much more understandable now.

maexled commented on src/realtime_metrics/main.py:
```diff
@@ -79,6 +76,11 @@ def run_analysis():
     
     for trip_id in trips.keys():
         trip_updates = trips[trip_id]
```
```suggestion
        trip_updates: list[tuple[TripUpdate, StopTimeUpdate]] = trips[trip_id]
```
To take the new type since trip_updates was already defined and thus uses another type (list[TripUpdate])

maexled commented on src/realtime_metrics/main.py:
```diff
@@ -347,8 +349,8 @@ def availability_acceptable_stop_time_updates(stop_time_updates: list[tuple[Trip
+    time_frame_end: end of the time frame to observe, in seconds since 1970
 
     Returns:
     A float containing the percentage of one minute slots with two or more updates.
```
Docstring seems not to be correclty formatted: 
![grafik](https://github.com/user-attachments/assets/afe6e157-1f23-4819-a64d-c90aa67e8e8a)

WuestFlorian replied:
I adapted the format.

maexled reviewed (commented):
the metric still returns 0 for me: Availability of acceptable stop time updates: 0.0%

Is this correct?

maexled commented:
And I agree with you to just use the time range from the first stop update timestamp to the last of each trip, I will do the same for VehiclePositions metrics

WuestFlorian commented:
> the metric still returns 0 for me: Availability of acceptable stop time updates: 0.0%
> 
> Is this correct?

Yes, this is correct. There is still only one update in each minute, so every slot has fewer than 2 updates. However, if you lower the required updates to 1,  the metric becomes almost 100%, as it should.

WuestFlorian committed:
Apply code review suggestion

Renamed a variable + comment, added type hints, adapted docstring format

1Maxnet1 commented on src/realtime_metrics/main.py:
```diff
@@ -362,21 +371,21 @@ def availability_acceptable_stop_time_updates(stop_time_updates: list[tuple[Trip
+        # get time in minutes by removing part of the timestamp responsible for seconds
+        time_rounded_in_minutes = int(trip_update.timestamp.replace(tzinfo=timezone.utc).timestamp() / 60) * 60
 
         # skip, if outide of time frame
```
Typo:
```suggestion
        # skip, if outside of time frame
```

1Maxnet1 reviewed (approved):
Looks goo dot me too, two more proposals which you could include?

WuestFlorian committed:
Code review suggestions

Reworked the converting of the timestamp to minutes to make it more understandable

WuestFlorian committed:
Merge branch 'main' into fix/stop-time-availability

WuestFlorian committed:
formatting
